airflow/utils/session.py

```diff
@@ -70,3 +70,81 @@ def wrapper(*args, **kwargs) -> RT:
                 return func(*args, session=session, **kwargs)
 
     return wrapper
+
+
+@provide_session
+@contextlib.contextmanager
+def create_global_lock(session=None, pg_lock_id=1, mysql_lock_name='init', mysql_lock_timeout=1800):
+    """Contextmanager that will create and teardown a global db lock."""
+    dialect = session.connection().dialect
+    if dialect.name == 'postgresql':
+        session.connection().execute('select PG_ADVISORY_LOCK(' + str(pg_lock_id) + ');')
+
+    if dialect.name == 'mysql' and dialect.server_version_info >= (
+        5,
+        6,
+    ):
+        session.connection().execute(
+            "select GET_LOCK('" + mysql_lock_name + "'," + str(mysql_lock_timeout) + ");"
+        )
+    try:
+        yield None
+    finally:
+        if dialect.name == 'postgresql':
+            session.connection().execute('select PG_ADVISORY_UNLOCK(' + str(pg_lock_id) + ');')
+
+        if dialect.name == 'mysql' and dialect.server_version_info >= (
+            5,
+            6,
+        ):
+            session.connection().execute("select RELEASE_LOCK('" + mysql_lock_name + "');")
+
+
+def provide_global_lock(func: Callable[..., RT]) -> Callable[..., RT]:
+    """
+    Function decorator that provides a global lock.
+    """
+    session_args_idx = find_session_idx(func)
+    func_params = signature(func).parameters
+
+    @wraps(func)
+    def wrapper(*args, **kwargs) -> RT:
+        pg_lock_id = __get_arg(args, kwargs, func_params, 'pg_lock_id', 1)
+        mysql_lock_name = __get_arg(args, kwargs, func_params, 'mysql_lock_name', 'init')
+        mysql_lock_timeout = __get_arg(args, kwargs, func_params, 'mysql_lock_timeout', 1800)
```

Comment:
Grgh, I really don’t like this, it’s over-complicating a function that’s only used a couple of times. Let’s scratch the decorator and use the more straightforward context manager approach?

Reply:
Done